File: core/server/search/score.go
package search

import (
	"sort"
	"strings"
)

// Match-quality tiers. Deliberately NOT BM25: FTS5 ranks weight terms by
// frequency within their own table's corpus, so scores from two packages are in
// different units and a perfect filename match can sort below a marginal mail
// hit. Match quality against visible text is unit-free and identical for every
// package.
//
// These values and the tiering below mirror core/lib/search/score.ts exactly.
// While both exist they are pinned to one another by a shared fixture
// (testdata/score-cases.json) asserted from both languages — two independent
// implementations of one ranking would otherwise diverge silently.
const (
	tierExactTitle      = 1000
	tierTitlePrefix     = 800
	tierAllTermsInTitle = 600
	tierTitleSubstring  = 400
	tierSecondaryMatch  = 200
	tierNoVisibleMatch  = 100
)

// normalize folds case and punctuation so 'Budget 2026' and 'budget-2026'
// compare equal. Punctuation becomes a space rather than being deleted, so
// 'budget-2026' and 'budget 2026' normalize identically instead of both
// collapsing to 'budget2026'.
func normalize(value string) string {
	var b strings.Builder
	b.Grow(len(value))
	lastSpace := true
	for _, r := range strings.ToLower(value) {
		isAlnum := (r >= 'a' && r <= 'z') || (r >= '0' && r <= '9')
		if isAlnum {
			b.WriteRune(r)
			lastSpace = false
			continue
		}
		if !lastSpace {
			b.WriteByte(' ')
			lastSpace = true
		}
	}
	return strings.TrimSpace(b.String())
}

// ScoreRow scores how well a row matches the query, using only text the row
// displays. A hit the backend matched on invisible content (a mail body, file
// contents) still scores, but below anything with a visible match.
func ScoreRow(include []string, row Row) int {
	terms := make([]string, 0, len(include))
	for _, t := range include {
		if n := normalize(t); n != "" {
			terms = append(terms, n)
		}
	}
	if len(terms) == 0 {
		return tierNoVisibleMatch
	}

	title := normalize(row.Title)
	query := strings.Join(terms, " ")

	if title == query {
		return tierExactTitle
	}
	if strings.HasPrefix(title, query) {
		return tierTitlePrefix
	}

	titleWords := strings.Split(title, " ")
	everyTermPrefixesAWord := true
	for _, term := range terms {
		found := false
		for _, word := range titleWords {
			if strings.HasPrefix(word, term) {
				found = true
				break
			}
		}
		if !found {
			everyTermPrefixesAWord = false
			break
		}
	}
	if everyTermPrefixesAWord {
		return tierAllTermsInTitle
	}

	if strings.Contains(title, query) {
		return tierTitleSubstring
	}

	var secondaryParts []string
	if row.Subtitle != "" {
		secondaryParts = append(secondaryParts, row.Subtitle)
	}
	if row.Meta != "" {
		secondaryParts = append(secondaryParts, row.Meta)
	}
	secondary := normalize(strings.Join(secondaryParts, " "))
	if secondary != "" {
		for _, term := range terms {
			if strings.Contains(secondary, term) {
				return tierSecondaryMatch
			}
		}
	}

	return tierNoVisibleMatch
}

// sortRows orders a merged cross-package slice in place. Tie-breaks, in order:
// score, then shorter title (a tighter match), then the package's Order — so
// the result does not depend on which source answered first.
func sortRows(rows []Row, include []string, orderBySlug map[string]int) {
	scores := make(map[string]int, len(rows))
	key := func(r Row) string { return r.Slug + "\x00" + r.ID }
	for _, r := range rows {
		scores[key(r)] = ScoreRow(include, r)
	}
	sort.SliceStable(rows, func(i, j int) bool {
		a, b := rows[i], rows[j]
		if sa, sb := scores[key(a)], scores[key(b)]; sa != sb {
			return sa > sb
		}
		if len(a.Title) != len(b.Title) {
			return len(a.Title) < len(b.Title)
		}
		return orderBySlug[a.Slug] < orderBySlug[b.Slug]
	})
}

File: core/lib/search/score.ts
import type { SearchRow } from './types'

const TIER_EXACT_TITLE = 1000
const TIER_TITLE_PREFIX = 800
const TIER_ALL_TERMS_IN_TITLE = 600
const TIER_TITLE_SUBSTRING = 400
const TIER_SECONDARY_MATCH = 200
const TIER_NO_VISIBLE_MATCH = 100

/**
 * Fold case and punctuation so 'Budget 2026' and 'budget-2026' compare equal.
 * Punctuation becomes a space rather than being deleted, so 'budget-2026' and
 * 'budget 2026' normalize identically instead of collapsing to 'budget2026'.
 */
function normalize(value: string): string {
    return value
        .toLowerCase()
        .replace(/[^a-z0-9]+/g, ' ')
        .trim()
}

/**
 * Score how well a row matches the query, using only text the row displays.
 *
 * Deliberately NOT BM25: FTS5 ranks weight terms by frequency within their own
 * table's corpus, so scores from two packages are in different units and a
 * perfect filename match can sort below a marginal mail hit. Match quality
 * against visible text is unit-free and identical for every package.
 */
export function scoreRow(includeTerms: string[], row: SearchRow): number {
    const terms = includeTerms.map(normalize).filter(Boolean)
    if (terms.length === 0) return TIER_NO_VISIBLE_MATCH

    const title = normalize(row.title)
    const query = terms.join(' ')

    if (title === query) return TIER_EXACT_TITLE
    if (title.startsWith(query)) return TIER_TITLE_PREFIX

    const titleWords = title.split(' ')
    const everyTermPrefixesAWord = terms.every(term =>
        titleWords.some(word => word.startsWith(term))
    )
    if (everyTermPrefixesAWord) return TIER_ALL_TERMS_IN_TITLE

    if (title.includes(query)) return TIER_TITLE_SUBSTRING

    const secondary = normalize([row.subtitle, row.meta].filter(Boolean).join(' '))
    if (secondary && terms.some(term => secondary.includes(term))) return TIER_SECONDARY_MATCH

    // The backend matched something we cannot see — a mail body or drive file
    // content. Keep it, but below anything with a visible match.
    return TIER_NO_VISIBLE_MATCH
}

/**
 * Sort comparator for a flat cross-package list. Tie-breaks, in order:
 * shorter title (a tighter match), then the package's nav.order so the
 * ordering is stable while results stream in from several packages.
 */
export function compareRows(
    a: SearchRow,
    b: SearchRow,
    includeTerms: string[],
    orderBySlug: Record<string, number>
): number {
    const scoreDelta = scoreRow(includeTerms, b) - scoreRow(includeTerms, a)
    if (scoreDelta !== 0) return scoreDelta

    const lengthDelta = a.title.length - b.title.length
    if (lengthDelta !== 0) return lengthDelta

    return (orderBySlug[a.slug] ?? 0) - (orderBySlug[b.slug] ?? 0)
}
